Replace server timeout { sleep until ... } loops with non-interrupting wait_until { ... } loops

SpComb opened the pull request:
Fixes #1986 to not interrupt moped operations.

This differs from the [agent's `WaitHelper`](https://github.com/kontena/kontena/blob/master/agent/lib/kontena/helpers/wait_helper.rb) in some ways:

* Renamed to `wait_until { condition }` to more closely match the `sleep interval until condition` that this is replacing
* The `wait_until!` method correctly returns the result
* The `wait_until` method only yields once if the interval is longer than the timeout
* The `wait_until` method does not yield if the timeout is zero
* The `wait_until` method takes the message as an optional first positional argument
* The debug and exception messages contains timing information
* More debug messages

This differs from the existing `Timeout::timeout` in some ways:

* Does not impose any timeout on the actual block operation - you need to use e.g. moped's own timeout mechanisms instead

jakolehm commented on server/app/models/distributed_lock.rb:
```diff
@@ -13,17 +15,16 @@ def self.with_lock(name, timeout = 60)
-        Timeout.timeout(timeout) do
-          sleep 0.01 until lock_id = self.obtain_lock(name)
-        end
+        lock_id = WaitHelper.wait(timeout: timeout) { self.obtain_lock(name) }
```
Why not include `WaitHelper` ?

SpComb replied:
Because this is a class method, so just `include WaitHelper` doesn't let you call `self.wait`

```
  1) DistributedLock.with_lock returns false if getting lock timeouts
     Failure/Error: described_class.with_lock('foo1') { sleep 0.1 }
     NoMethodError:
       undefined method `wait' for DistributedLock:Class
       Did you mean?  with
     # ./app/models/distributed_lock.rb:19:in `with_lock'
     # ./spec/models/distributed_lock_spec.rb:27:in `block (4 levels) in <top (required)>'
```

Is there a better way to get this working?

SpComb replied:
This changes the default `sleep` interval from 0.01 to 0.5. Looking at the calls to `with_dlock` in the server:

```ruby
    with_dlock("check_deploy_queue:#{service_deploy.grid_service_id}", 10) do
    with_dlock(DLOCK_KEY, nil) do
    with_dlock('mongodb_migrate', 60) do
```

I don't suppose this makes tooooooo much of a difference?

SpComb replied:
Might still be a good idea to bump this down to 0.05 or something, just to keep it as similar as possible.

jnummelin commented:
There is one missed `Timeout::timeout` call in `GridCertificates::GetCertificate` mutation

jnummelin commented on server/app/helpers/wait_helper.rb:
```diff
@@ -0,0 +1,51 @@
+      if value
+        return value
+      else
+        debug "wait... #{message}"
```
Should this be guarded with `if message`? Seeing just random `wait...` entries in the log might be confusing.

SpComb replied:
They will have the including class name for context.

```
D, [2017-03-16T09:08:49.254409 #27900] DEBUG -- StackRemoveWorker: wait... 
```

jnummelin replied:
Yes, but it still gives no idea what it's waiting for. It's on `DEBUG` so should not matter that much. :)

SpComb replied:
Cherry-picked from  #1988

SpComb commented:
> There is one missed Timeout::timeout call in GridCertificates::GetCertificate mutation

It doesn't do any mongo operations AFAIK, so I didn't touch it... it looks like the `acme_client` configures a timeout, so this could be replaced with `wait`, but then I would need to test it... ain't got no time for that.

SpComb commented:
>> There is one missed Timeout::timeout call in GridCertificates::GetCertificate mutation
> 
> It doesn't do any mongo operations AFAIK, so I didn't touch it... it looks like the acme_client configures a timeout, so this could be replaced with wait, but then I would need to test it... ain't got no time for that.

Cherry-picked from  #1988

SpComb commented:
>>> There is one missed Timeout::timeout call in GridCertificates::GetCertificate mutation
>>
>>   It doesn't do any mongo operations AFAIK, so I didn't touch it... it looks like the acme_client configures a timeout, so this could be replaced with wait, but then I would need to test it... ain't got no time for that.
> 
> Cherry-picked from #1988

The replacement wait was broken, un-pushed this bit. Let's just leave it as it is.

SpComb commented:
Renamed `wait` -> `wait_until`

SpComb commented on server/app/helpers/wait_helper.rb:
```diff
@@ -28,7 +28,7 @@ def wait(timeout: WAIT_TIMEOUT, interval: WAIT_INTERVAL, message: nil, &block)
       if value
         return value
       else
-        debug "wait... #{message}" if message
```
Eh... somehow accidentially reverted this change. Must have still had it open in my editor.

jnummelin commented on server/app/helpers/wait_helper.rb:
```diff
@@ -0,0 +1,51 @@
+    if value = wait_until(**opts, &block)
+      return value
+    else
+      raise Timeout::Error
```
Would it make sense to put in some "context" for the error? Now the logs don't show much on these:
```
E, [2017-03-16T09:45:47.758033 #16] ERROR -- GridServiceInstanceDeployer: Timeout::Error
api_1      | E, [2017-03-16T09:45:47.758132 #16] ERROR -- GridServiceInstanceDeployer: /app/app/helpers/wait_helper.rb:45:in `wait_until!'
api_1      | /app/app/services/grid_service_instance_deployer.rb:36:in `wait_for_service_to_start'
api_1      | /app/app/services/grid_service_instance_deployer.rb:22:in `deploy'
api_1      | /app/app/services/grid_service_deployer.rb:112:in `block in deploy_service_instance'
api_1      | /usr/lib/ruby/gems/2.3.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `public_send'
```
It would help debugging if there would be something like:
```
E, [2017-03-16T09:45:47.758033 #16] ERROR -- GridServiceInstanceDeployer: Timeout::Error: service startup
```

SpComb replied:
AFAIK the `GridServiceInstanceDeployer: Timeout::Error` is equivalent to what it currently logs when using `Timeout::timeout { ... }`

Yes, this could be improved to log something like `Timeout after 300s while waiting until service is running` if each of the callers passes in a `message: ...`.

jnummelin replied:
Yes, pull in the provided message somehow if given

SpComb commented:
Changed the `wait_until(message)` to be an optional positional parameter, similar to e.g. https://github.com/rubytools/waitutil

Added messages to all the server waits:

```
D, [2017-03-16T14:14:28.474436 #22752] DEBUG -- StackRemoveWorker: waiting 532.8s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:28.976228 #22752] DEBUG -- StackRemoveWorker: waiting 533.3s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:29.477632 #22752] DEBUG -- StackRemoveWorker: waiting 533.8s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:29.979917 #22752] DEBUG -- StackRemoveWorker: waiting 534.4s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:30.481552 #22752] DEBUG -- StackRemoveWorker: waiting 534.9s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:30.986080 #22752] DEBUG -- StackRemoveWorker: waiting 535.4s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:31.488383 #22752] DEBUG -- StackRemoveWorker: waiting 535.9s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:31.992230 #22752] DEBUG -- StackRemoveWorker: waiting 536.4s of 600.0s until: service development/redis/redis is gone
D, [2017-03-16T14:14:32.496606 #22752] DEBUG -- StackRemoveWorker: waiting 536.9s of 600.0s until: service development/redis/redis is gone
```

SpComb commented on server/app/jobs/grid_service_remove_worker.rb:
```diff
@@ -21,18 +22,19 @@ def remove_grid_service(grid_service)
       grid_service.destroy
-    rescue Timeout::Error
-      error "service remove timed out #{grid_service.to_path}"
+    rescue => error
```
Also catching things like RPC errors now... although irrelevant after #1873

SpComb commented:
See #1991 for the zero-timeout edgecase breaking the `GridServiceRemoveWorker`.
